crates/artifact_content/src/file_system_state.rs

```diff
@@ -0,0 +1,530 @@
+            Some(type_and_field) => {
+                self.nested_files
+                    .entry(type_and_field.parent_object_entity_name)
+                    .or_default()
+                    .entry(type_and_field.selectable_name)
+                    .or_default()
+                    .insert(path_and_content.artifact_path.file_name, value);
+            }
+            None => {
+                self.root_files
+                    .insert(path_and_content.artifact_path.file_name, value);
+            }
+        }
+    }
+
+    pub fn from_artifacts(artifacts: Vec<ArtifactPathAndContent>) -> Self {
+        let mut state = Self::default();
+        for artifact in artifacts {
+            state.insert(&artifact);
+        }
+        state
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.root_files.is_empty() && self.nested_files.is_empty()
+    }
+
+    // Computes filesystem operations needed to transform this state into the target state.
+    // Returns operations to create, update, and delete files and directories. Files are updated
+    // only when their hash differs. If the current state is empty, emits a DeleteDirectory
+    // operation for the artifact root followed by recreation of all files. Empty directories
+    // are automatically removed after their files are deleted.
+    pub fn diff(&self, new: &Self, artifact_directory: &PathBuf) -> Vec<FileSystemOperation> {
+        let mut operations: Vec<FileSystemOperation> = Vec::new();
+
+        if self.is_empty() {
+            operations.push(FileSystemOperation::DeleteDirectory(
+                artifact_directory.to_path_buf(),
+            ));
+
+            for (server_object, selectables) in &new.nested_files {
+                let server_object_path = artifact_directory.join(server_object.to_string());
+                operations.push(FileSystemOperation::CreateDirectory(
+                    server_object_path.clone(),
+                ));
+
+                for (selectable, files) in selectables {
+                    let selectable_path = server_object_path.join(selectable.to_string());
+                    operations.push(FileSystemOperation::CreateDirectory(
+                        selectable_path.clone(),
+                    ));
+
+                    for (file_name, (content, _)) in files {
+                        let file_path = selectable_path.join(file_name.to_string());
+                        operations.push(FileSystemOperation::WriteFile(file_path, content.clone()));
+                    }
+                }
+
+                for (file_name, (content, _)) in &new.root_files {
+                    let file_path = artifact_directory.join(file_name.to_string());
+                    operations.push(FileSystemOperation::WriteFile(file_path, content.clone()));
+                }
+            }
+
+            return operations;
+        }
+
+        let mut new_server_objects: HashSet<&ServerObjectEntityName> = HashSet::new();
```

Comment:
We should not generally create references to `&Copy` types, and anything that's been interned (e.g. `ServerObjectEntityName` and `SelectableName`) implements Copy. So instead, let's just do `*server_object_entity_name` wherever necessary